### PR TITLE
Deploy crd

### PR DESCRIPTION
**Pull request by Tal-or**
Deploy/Undeploy CRD

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -5,6 +5,7 @@ go 1.16
 require (
 	github.com/drone/envsubst v1.0.3
 	github.com/spf13/cobra v1.1.1
+	github.com/spf13/pflag v1.0.5
```
Comment (fromanirh) [inline go.mod:8]: why is this needed?

File: pkg/commands/consts.go [hunk 2]
```diff
@@ -0,0 +1,3 @@
+package commands
+
+const crdDefaultManifestPath = "manifests/api/crd.yaml"
```
Comment (fromanirh) [inline pkg/commands/consts.go:3]: this should be in the `manifests` package

File: pkg/commands/deploy.go [hunk 3]
```diff
@@ -17,35 +17,107 @@
...
 	}
+	opts.AddFlags(remove.Flags())
 	return remove
 }
+
+func (d *deployOptions) AddFlags(fs *pflag.FlagSet) {
+	fs.StringVar(&d.crdPath, "crd-manifest", crdDefaultManifestPath, "crd manifest path")
+}
+
+func (d *deployOptions) Validate() error {
```
Comment (fromanirh) [inline pkg/commands/deploy.go:82]: let's not validate: let's just try to read and bail out if it fails.

File: go.mod (hunk 1 above)
Comment (Tal-or) [reply go.mod:8]: Reading the command flags from the user

Comment (fromanirh) [reply go.mod:8]: I'm not sure it's actually needed, let's check

File: pkg/commands/deploy.go [hunk 4]
```diff
@@ -17,35 +17,107 @@
...
+			}
+
+			if err := opts.Remove(); err != nil {
+				return err
+			}
 			return nil
 		},
 		Args: cobra.NoArgs,
 	}
+	opts.AddFlags(remove.Flags())
```
Comment (fromanirh) [inline pkg/commands/deploy.go:74]: what happens if we don't `AddFlags` explicitely?

Comment (fromanirh) [reply pkg/commands/deploy.go:74]: like, what if we do like https://github.com/fromanirh/deployer/blob/main/pkg/commands/validate.go#L44

Comment (Tal-or) [reply pkg/commands/deploy.go:74]: I don't know if it's possible let me check

File: pkg/commands/deploy.go [hunk 5]
```diff
@@ -17,35 +17,107 @@
...
+}
+
+func (d *deployOptions) Validate() error {
+	if _, err := ioutil.ReadFile(d.crdPath); errors.Is(err, fs.ErrNotExist) {
+		return fmt.Errorf("crd manifest path %s not exist", d.crdPath)
+	}
+	return nil
+}
+
+func (d *deployOptions) Deploy() error {
```
Comment (fromanirh) [inline pkg/commands/deploy.go:89]: this should be a free function if possible:
```
func  Deploy(opts deployOptions) error
```

File: pkg/commands/deploy.go [hunk 6]
```diff
@@ -17,35 +17,107 @@
...
+	}
+
+	err = cs.Create(context.TODO(), crd)
+	if err !=nil {
+		return err
+	}
+	return nil
+}
+
+func (d *deployOptions) Remove() error {
```
Comment (fromanirh) [inline pkg/commands/deploy.go:107]: this should be a free function if possible:
```
func  Remove(opts deployOptions) error
```

File: pkg/commands/render.go [hunk 7]
```diff
@@ -48,7 +48,7 @@ func NewRenderCommand(commonOpts *CommonOptions) *cobra.Command {
 func renderManifests(cmd *cobra.Command, commonOpts *CommonOptions, opts *renderOptions, args []string) error {
 	var objs []runtime.Object
 
-	crd, err := manifests.LoadAPICRD()
+	crd, err := manifests.LoadAPICRD(manifests.CrdDefaultManifestPath)
```
Comment (fromanirh) [inline pkg/commands/render.go:51]: this is a good idea but we need to push it back. The fact we load the manifests is an implementation detail that the consumers of the `manifests` package should not concern with. I'm going to push a rename dropping the `Load` prefix to emphasize this: the `manifests` package will give the client the object, doesn't matter how it gets it.

Later, past MVP, we will add an option to possibly override the builtin manifests.

Comment (Tal-or) [reply pkg/commands/render.go:51]: So basically the client shouldn't be able to provide the path? 
If that's the case i need to remove also the path flag from the command

Comment (fromanirh) [reply pkg/commands/render.go:51]: yes, not now. We need to postpone this feature.

File: pkg/commands/deploy.go [hunk 8]
```diff
@@ -37,15 +42,53 @@ func NewDeployCommand(commonOpts *CommonOptions) *cobra.Command {
...
+				return err
+			}
 			return nil
 		},
 		Args: cobra.NoArgs,
 	}
 	return remove
 }
+
+func Deploy(opts *deployOptions) error {
```
Comment (fromanirh) [inline pkg/commands/deploy.go:60]: we will need to reshape this code a bit when the ongoing refactoring (not even a PR posted) of the `deployer` package lands, but for now looks good enough.

There is no logging/explicit feedback of the operations, but we can fix in a later PR.

File: pkg/manifests/manifests.go [hunk 9]
```diff
@@ -36,6 +36,7 @@ const (
 	ComponentAPI                      = "api"
 	ComponentSchedulerPlugin          = "sched"
 	ComponentResourceTopologyExporter = "rte"
+	CrdDefaultManifestPath = "manifests/api/crd.yaml"
```
Comment (fromanirh) [inline pkg/manifests/manifests.go:39]: the paths should be all changed in a separate PR, but we can fix later.

**Review by fromanirh — APPROVED** [review]

**Comment by fromanirh** [comment]
please address the conflicts - if possible, revert the addition of the constant for the CRD path.

**Comment by Tal-or** [comment]
I think you need to explain to me how you vision this whole deployer thingy. 
For example what you mean by "refactoring (not even a PR posted) of the deployer package lands"
It can save us some review cycles.

**Review by fromanirh — APPROVED** [review]

**Comment by fromanirh** [comment]
> I think you need to explain to me how you vision this whole deployer thingy.
> For example what you mean by "refactoring (not even a PR posted) of the deployer package lands"
> It might save us some review cycles.

Yeah, the problem is we design as we go, in this stage. But yes, let's sync more offline.
